### PR TITLE
use TLSConfig.ServerName as the client hostname if provided

### PR DESCRIPTION
**Pull request by marten-seemann**
For some reason, Github didn't let me push into PR #686, so here's a new PR.

### CODE REVIEW

**Comment by marten-seemann** [comment]
Hi @phuslu,

thanks for your PR! I added some tests and removed some now obsolete code in the crypto package, and we can now merge these changes.
Apparently, @googlebot is not happy that I created a new branch for that. Can you please confirm here that you're ok with us merging your contribution here in this PR?

File: crypto/cert_manager.go [hunk 1]
```diff
@@ -106,8 +106,8 @@ func (c *certManager) Verify(hostname string) error {
 
 	var opts x509.VerifyOptions
 	if c.config != nil {
+		// we don't need to care about the tls.Config.ServerName here, since hostname has already been set to that value in the session setup
```
Comment (lucas-clemente) [inline crypto/cert_manager.go]: I still don't really understand why you are doing this change… If the config is not nil, you don't set DNSName at all, not even to hostname, right? How is the verification supposed to work then?

**Comment by phuslu** [comment]
Hi @marten-seemann and others, I'm OK for that.

File: crypto/cert_manager.go (hunk 1 above)
Comment (marten-seemann) [inline crypto/cert_manager.go]: Oh, right. Must have missed that. Please review again.

**Pull request merged**